File: src/main/java/org/basex/server/trigger/BaseXClient.cs
/*
 * Language Binding for BaseX.
 * Works with BaseX 6.3.1 and later
 * Documentation: http://basex.org/api
 *
 * (C) BaseX Team 2005-11, BSD License
 */
using System;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Generic;
using System.IO;
using System.Threading;

namespace BaseXClient
{
  class Session
  {
    private byte[] cache = new byte[4096];
    public NetworkStream stream;
    private TcpClient socket;
    private string info = "";
    private string shost;
    private int bpos;
    private int bsize;
	private TcpClient tsocket;
	private NetworkStream tstream;
	private Dictionary<string, TriggerNotification> tn;
	private int tport = 1985;

    /** see readme.txt */
    public Session(string host, int port, string username, string pw)
    {
      socket = new TcpClient(host, port);
      stream = socket.GetStream();
	  shost = host;
      string ts = Receive();
      Send(username);
      Send(MD5(MD5(pw) + ts));
      if (stream.ReadByte() != 0)
      {
        throw new IOException("Access denied.");
      }
    }

    /** see readme.txt */
    public void Execute(string com, Stream ms)
    {
      Send(com);
      Init();
      Receive(ms);
      info = Receive();
      if(!Ok())
      {
        throw new IOException(info);
      }
    }

    /** see readme.txt */
    public String Execute(string com)
    {
      MemoryStream ms = new MemoryStream();
      Execute(com, ms);
      return System.Text.Encoding.UTF8.GetString(ms.ToArray());
    }

    /** see readme.txt */
    public Query Query(string q)
    {
      return new Query(this, q);
    }

    /** see readme.txt */
    public void Create(string name, Stream s)
    {
      stream.WriteByte(8);
      Send(name);
      while (true)
      {
      	int t = s.ReadByte();
      	if (t == -1) break;
      	stream.WriteByte(Convert.ToByte(t));
      }
      stream.WriteByte(0);
      info = Receive();
      if(!Ok())
      {
        throw new IOException(info);
      }
    }
    
    /** see readme.txt */
    public void Add(string name, string target, Stream s)
    {
      stream.WriteByte(9);
      Send(name);
      Send(target);
      while (true)
      {
      	int t = s.ReadByte();
      	if (t == -1) break;
      	stream.WriteByte(Convert.ToByte(t));
      }
      stream.WriteByte(0);
      info = Receive();
      if(!Ok())
      {
        throw new IOException(info);
      }
    }
	
	/** see readme.txt */
  	public void AttachTrigger(string name, TriggerNotification notify)
	{
	  stream.WriteByte(10);
	  Send(name);
	  if(tsocket == null)
	  {	
		tn = new Dictionary<string, TriggerNotification>();
		tsocket = new TcpClient(shost, tport);
		tstream = tsocket.GetStream();
		string id = Receive();
		byte[] msg = System.Text.Encoding.UTF8.GetBytes(id);
      	tstream.Write(msg, 0, msg.Length);
      	tstream.WriteByte(0);
	    Thread t = new Thread(Listen);
		t.Start();
	  }
	  info = Receive();
	  if(!Ok())
      {
        throw new IOException(info);
      }
	  tn.Add(name, notify);
	}
	
	/** Listens to trigger socket */
	private void Listen() 
	{
	  while (true)
	    {
		  String name = readS();
		  String val = readS();
		  tn[name].Update(val);
		}
	}
	
	/** Returns trigger message */
	private string readS() 
	  {
		MemoryStream ms = new MemoryStream();
	  	while (true) 
		  {
			int b = tstream.ReadByte();
			if (b == 0) break;
			ms.WriteByte((byte) b);
		  }
		return System.Text.Encoding.UTF8.GetString(ms.ToArray()); 
	  }
	
	/** see readme.txt */
	public void Trigger(string query, string name, string msg, string mode)
	{
	  Execute("xquery db:trigger(" + query + ", " + name + ", " + notification + "," + mode")");
	}
	
	/** see readme.txt */
	public void DetachTrigger(string name)
	{
	  stream.WriteByte(11);
	  Send(name);
	  info = Receive();
	  if(!Ok())
      {
        throw new IOException(info);
      }
	  tn.Remove(name);
	}

    /** see readme.txt */
    public string Info
    {
      get
      {
        return info;
      }
    }

    /** see readme.txt */
    public void Close()
    {
      Send("exit");
      socket.Close();
    }

    /** Initializes the byte transfer. */
    private void Init()
    {
      bpos = 0;
      bsize = 0;
    }

    /** Returns a single byte from the socket. */
    private byte Read()
    {
      if (bpos == bsize)
      {
        bsize = stream.Read(cache, 0, 4096);
        bpos = 0;
      }
      return cache[bpos++];
    }

    /** Receives a string from the socket. */
    private void Receive(Stream ms)
    {
      while (true)
      {
        byte b = Read();
        if (b == 0) break;
        ms.WriteByte(b);
      }
    }

    /** Receives a string from the socket. */
    public string Receive()
    {
      MemoryStream ms = new MemoryStream();
      Receive(ms);
      return System.Text.Encoding.UTF8.GetString(ms.ToArray());
    }

    /** Sends strings to server. */
    public void Send(string message)
    {
      byte[] msg = System.Text.Encoding.UTF8.GetBytes(message);
      stream.Write(msg, 0, msg.Length);
      stream.WriteByte(0);
    }

    /** Returns success check. */
    public bool Ok()
    {
      return Read() == 0;
    }

    /** Returns the md5 hash of a string. */
    private string MD5(string input)
    {
      MD5CryptoServiceProvider MD5 = new MD5CryptoServiceProvider();
      byte[] hash = MD5.ComputeHash(Encoding.UTF8.GetBytes(input));

      StringBuilder sb = new StringBuilder();
      foreach (byte h in hash)
      {
        sb.Append(h.ToString("x2"));
      }
      return sb.ToString();
    }
  }

  class Query
  {
    private Session session;
    private string id;
    private string next;

    /** see readme.txt */
    public Query(Session s, string query)
    {
      session = s;
      id = Exec(0, query);
    }

    /** see readme.txt */
    public string Init()
    {
      return Exec(4, id);
    }

    /** see readme.txt */
    public void Bind(string name, string value)
    {
      Exec(3, id + '\0' + name + '\0' + value + '\0');
    }

    /** see readme.txt */
    public bool More()
    {
      next = Exec(1, id);
      return next.Length != 0;
    }

    /** see readme.txt */
    public string Next()
    {
      return next;
    }

    /** see readme.txt */
    public string Execute()
    {
      return Exec(5, id);
    }

    /** see readme.txt */
    public string Info()
    {
      return Exec(6, id);
    }

    /** see readme.txt */
    public string Close()
    {
      return Exec(2, id);
    }

    /** see readme.txt */
    private string Exec(byte cmd, string arg)
    {
      session.stream.WriteByte(cmd);
      session.Send(arg);
      string s = session.Receive();
      if(!session.Ok())
      {
        throw new IOException(session.Receive());
      }
      return s;
    }
  }
	
  interface TriggerNotification 
	{
		void Update(string data);
	}
}

File: src/main/java/BaseXClient.java
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

/**
 * Java client for BaseX.
 * Works with BaseX 6.3.1 and later
 * Documentation: http://basex.org/api
 *
 * (C) BaseX Team 2005-11, BSD License
 */
public final class BaseXClient {
  /** Event notifications. */
  final Map<String, EventNotifier> notifiers =
    new HashMap<String, EventNotifier>();
  /** Output stream. */
  final OutputStream out;
  /** Socket. */
  private final Socket socket;
  /** Cache. */
  private final BufferedInputStream in;
  /** Command info. */
  private String info;
  /** Socket event reference. */
  Socket esocket;
  /** Socket host name. */
  String ehost;

  /**
   * Constructor.
   * @param host server name
   * @param port server port
   * @param usern user name
   * @param pw password
   * @throws IOException Exception
   */
  public BaseXClient(final String host, final int port, final String usern,
      final String pw) throws IOException {
    socket = new Socket();
    socket.connect(new InetSocketAddress(host, port), 5000);
    in = new BufferedInputStream(socket.getInputStream());
    out = socket.getOutputStream();
    ehost = host;
    // receive timestamp
    final String ts = receive();
    // send {Username}0 and hashed {Password/Timestamp}0
    send(usern);
    send(md5(md5(pw) + ts));

    // receive success flag
    if(!ok()) throw new IOException("Access denied.");
  }

  /**
   * Executes a command and serializes the result to an output stream.
   * @param cmd command
   * @param o output stream
   * @throws IOException Exception
   */
  public void execute(final String cmd, final OutputStream o)
      throws IOException {
    // send {Command}0
    send(cmd);
    receive(o);
    info = receive();
    if(!ok()) throw new IOException(info);
  }

  /**
   * Executes a command and returns the result.
   * @param cmd command
   * @return result
   * @throws IOException Exception
   */
  public String execute(final String cmd) throws IOException {
    final ByteArrayOutputStream os = new ByteArrayOutputStream();
    execute(cmd, os);
    return os.toString("UTF-8");
  }

  /**
   * Creates a query object.
   * @param query query string
   * @return query
   * @throws IOException Exception
   */
  public Query query(final String query) throws IOException {
    return new Query(query);
  }

  /**
   * Creates a database.
   * @param name name of database
   * @param input xml input
   * @throws IOException I/O exception
   */
  public void create(final String name, final InputStream input)
      throws IOException {
    out.write(8);
    send(name);
    send(input);
  }

  /**
   * Adds a database.
   * @param name name of document
   * @param target target path
   * @param input xml input
   * @throws IOException I/O exception
   */
  public void add(final String name, final String target,
      final InputStream input) throws IOException {
    out.write(9);
    send(name);
    send(target);
    send(input);
  }

  /**
   * Watches an event.
   * @param name event name
   * @param notifier event notification
   * @throws IOException I/O exception
   */
  public void watch(final String name,
      final EventNotifier notifier) throws IOException {
    out.write(10);
    send(name);
    if(esocket == null) {
      final int eport = Integer.parseInt(receive());
      // initialize event socket
      esocket = new Socket();
      esocket.connect(new InetSocketAddress(ehost, eport), 5000);
      final OutputStream os = esocket.getOutputStream();
      receive(in, os);
      os.write(0);
      os.flush();
      listen(esocket.getInputStream());
    }
    info = receive();
    if(!ok()) throw new IOException(info);
    notifiers.put(name, notifier);
  }

  /**
   * Starts the listener thread.
   * @param is input stream
   */
  private void listen(final InputStream is) {
    new Thread() {
      @Override
      public void run() {
        try {
          while(true) {
            final BufferedInputStream bi = new BufferedInputStream(is);
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            receive(bi, baos);
            final String name = baos.toString("UTF-8");
            baos = new ByteArrayOutputStream();
            receive(bi, baos);
            final String data = baos.toString("UTF-8");
            notifiers.get(name).notify(data);
          }
        } catch(final Exception ex) { }
      }
    }.start();
  }

  /**
   * Unwatches an event.
   * @param name event name
   * @throws IOException I/O exception
   */
  public void unwatch(final String name) throws IOException {
    out.write(11);
    send(name);
    info = receive();
    if(!ok()) throw new IOException(info);
    notifiers.remove(name);
  }

  /**
   * Sends an input stream to the server.
   * @param input xml input
   * @throws IOException I/O exception
   */
  private void send(final InputStream input) throws IOException {
    int l;
    final BufferedInputStream bis = new BufferedInputStream(input);
    final BufferedOutputStream bos = new BufferedOutputStream(out);
    while((l = bis.read()) != -1) bos.write(l);
    bos.write(0);
    bos.flush();
    info = receive();
    if(!ok()) throw new IOException(info);
  }

  /**
   * Returns command information.
   * @return string info
   */
  public String info() {
    return info;
  }

  /**
   * Closes the session.
   * @throws IOException Exception
   */
  public void close() throws IOException {
    send("exit");
    out.flush();
    if(esocket != null) esocket.close();
    socket.close();
  }

  /**
   * Checks the next success flag.
   * @return value of check
   * @throws IOException Exception
   */
  boolean ok() throws IOException {
    out.flush();
    return in.read() == 0;
  }

  /**
   * Returns the next received string.
   * @return String result or info
   * @throws IOException I/O exception
   */
  String receive() throws IOException {
    final ByteArrayOutputStream os = new ByteArrayOutputStream();
    receive(os);
    return os.toString("UTF-8");
  }

  /**
   * Sends a string to the server.
   * @param s string to be sent
   * @throws IOException I/O exception
   */
  void send(final String s) throws IOException {
    out.write((s + '\0').getBytes("UTF8"));
  }

  /**
   * Receives a string and writes it to the specified output stream.
   * @param bis input stream
   * @param o output stream
   * @throws IOException I/O exception
   */
  void receive(final BufferedInputStream bis,
      final OutputStream o) throws IOException {
    while(true) {
      final int b = bis.read();
      if(b == 0 || b == -1) break;
      o.write(b);
    }
  }

  /**
   * Receives a string and writes it to the specified output stream.
   * @param o output stream
   * @throws IOException I/O exception
   */
  private void receive(final OutputStream o) throws IOException {
    out.flush();
    while(true) {
      final int b = in.read();
      if(b == 0 || b == -1) break;
      o.write(b);
    }
  }

  /**
   * Returns an MD5 hash.
   * @param pw String
   * @return String
   */
  private static String md5(final String pw) {
    final StringBuilder sb = new StringBuilder();
    try {
      final MessageDigest md = MessageDigest.getInstance("MD5");
      md.update(pw.getBytes());
      for(final byte b : md.digest()) {
        final String s = Integer.toHexString(b & 0xFF);
        if(s.length() == 1) sb.append('0');
        sb.append(s);
      }
    } catch(final NoSuchAlgorithmException ex) {
      // should not occur
      ex.printStackTrace();
    }
    return sb.toString();
  }

  /**
   * Inner class for iterative query execution.
   */
  public final class Query {
    /** Query id. */
    private final String id;
    /** Next result item. */
    private String next;

    /**
     * Standard constructor.
     * @param query query string
     * @throws IOException I/O exception
     */
    public Query(final String query) throws IOException {
      id = exec(0, query);
    }

    /**
     * Initializes the query.
     * @return result header
     * @throws IOException I/O exception
     */
    public String init() throws IOException {
      return exec(4, id);
    }

    /**
     * Binds a variable.
     * @param name name of variable
     * @param value value
     * @throws IOException I/O exception
     */
    public void bind(final String name, final String value)
        throws IOException {
      exec(3, id + '\0' + name + '\0' + value + '\0');
    }

    /**
     * Checks for the next item.
     * @return result of check
     * @throws IOException I/O exception
     */
    public boolean more() throws IOException {
      next = exec(1, id);
      return next.length() != 0;
    }

    /**
     * Returns the next item.
     * @return item string
     */
    public String next() {
      return next;
    }

    /**
     * Returns the whole result of the query.
     * @return query result
     * @throws IOException I/O Exception
     */
    public String execute() throws IOException {
      return exec(5, id);
    }

    /**
     * Returns the query info.
     * @return query info
     * @throws IOException I/O exception
     */
    public String info() throws IOException {
      return exec(6, id);
    }

    /**
     * Closes the query.
     * @return result footer
     * @throws IOException I/O exception
     */
    public String close() throws IOException {
      final String s = exec(2, id);
      out.flush();
      return s;
    }

    /**
     * Executes the specified command.
     * @param cmd command
     * @param arg argument
     * @return resulting string
     * @throws IOException I/O exception
     */
    private String exec(final int cmd, final String arg)
        throws IOException {
      out.write(cmd);
      send(arg);
      final String s = receive();
      if(!ok()) throw new IOException(receive());
      return s;
    }
  }

  /**
   * Interface for event notifications.
   */
  public interface EventNotifier {
    /**
     * Invoked when a database event was fired.
     * @param value event string
     */
    void notify(final String value);
  }
}
